File: generator/prov/megajack/fizz.lua

local t = { -- champagne buttles table
	10, 10, 20, 150, 150,
}

local m = 0
for i = 1,#t do
	m = m + t[i]
end
m = m / #t

local M = 0
local n = 0
for i = 1,#t do
	for j = i+1,#t do
		if t[i] == t[j] then
			M = M + t[i]*4
		else
			M = M + t[i]+t[j]
		end
		n = n + 1
	end
end
M = M / n
print(string.format("len = %d, count = %d, avr bottle gain = %g, M = %g", #t, n, m, M))

File: game/slot/megajack/champagne/fizz.go
package champagne

import (
	"math/rand/v2"
)

// len = 5, count = 10, avr bottle gain = 68, M = 168
var Bottles = [...]float64{
	10, 10, 20, 150, 150,
}

type WinBottle struct {
	Mult float64 `json:"mult" yaml:"mult" xml:"mult,attr"` // bet multiplier
	Pay  float64 `json:"pay" yaml:"pay" xml:"pay,attr"`    // pay by this cell
}

func ChampagneSpawn(bet float64) (any, float64) {
	var res [5]WinBottle
	var cash float64

	var p = Bottles
	rand.Shuffle(len(p), func(i, j int) {
		p[i], p[j] = p[j], p[i]
	})
	for i := range res {
		res[i].Mult = p[i]
		res[i].Pay = bet * p[i]
	}
	cash = p[0] + p[1]
	if p[0] == p[1] {
		cash *= 2
	}
	return res, bet * cash
}
